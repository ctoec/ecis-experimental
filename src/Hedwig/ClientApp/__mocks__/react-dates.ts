const DayPickerSingleDateController = ({ ...props }) =>
	`<DayPickerSingleDateController" ${JSON.stringify(props)}/>`;

export { DayPickerSingleDateController };
