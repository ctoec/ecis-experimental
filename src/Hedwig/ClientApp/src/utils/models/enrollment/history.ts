import { ProcessStepProps } from '../../../components/ProcessList/ProcessStep';
import { DeepNonUndefineable, DeepNonUndefineableArray } from '../../types';
import {
	Enrollment,
	Funding,
	FundingSource,
	FamilyDetermination,
	C4KCertificate,
} from '../../../generated';
import { prettyAge, prettyFundingTime, NO_FUNDING } from '..';
import dateFormatter from '../../dateFormatter';
import { dateSorter } from '../../dateSorter';
import moment from 'moment';
import { c4kCertificateSorter } from '../c4kCertificate';
import { fundingStartSorter, getFundingTime } from '../funding';

// sortable process step props include:
// - sortDate, the field to use for sorting
// - sortWeight, a number to use as a sorting tiebreaker to ensure enrollment starts come before funding/determination steps,
// 	and enrollment ends come after
type SortableProcessStepProps = ProcessStepProps & { sortDate: Date | null; sortWeight: number };

/**
 * Generates SortableProcessStepProps array for a given array of enrollments,
 * also creating step props for fundings and family determinations
 * @param enrollment
 * @returns sorted SortableProcessStepProps array
 */
export function getEnrollmentHistoryProps(
	currentEnrollment: DeepNonUndefineable<Enrollment>
): ProcessStepProps[] {
	// allEnrollments is sorted in descending order by Entry
	// (pastEnrollments are sorted thusly in API response)
	var allEnrollments = [
		currentEnrollment,
		...(currentEnrollment.pastEnrollments ? currentEnrollment.pastEnrollments : []),
	];

	var processStepProps: SortableProcessStepProps[] = [];

	allEnrollments.forEach((enrollment, idx) => {
		processStepProps = processStepProps.concat(
			enrollmentStepProps(enrollment, idx, allEnrollments.length)
		);

		// funding steps
		processStepProps = processStepProps.concat(
			fundingStepProps(enrollment.fundings, enrollment.exit)
		);

		// only add determination steps and C4K funding steps for the current enrollment (others will be dupes)
		if (idx === 0) {
			// determination steps
			processStepProps = processStepProps.concat(
				determinationStepProps(enrollment.child.family.determinations)
			);

			// C4K certification steps
			processStepProps = processStepProps.concat(
				c4kCertificateStepProps(enrollment.child.c4KCertificates)
			);
		}
	});

	// return all steps sorted by `sortDate`, and sortWeight
	return processStepProps.sort(enrollmentHistoryPropsSorter);
}

export function enrollmentStepProps(
	_enrollment: DeepNonUndefineable<Enrollment>,
	enrollmentIndex: number,
	allEnrollmentsCount: number
) {
	var processStepProps: SortableProcessStepProps[] = [];

	// enrollment start step
	if (_enrollment.entry) {
		processStepProps.push(
			getSortableStep({
				heading:
					enrollmentIndex === allEnrollmentsCount - 1
						? // heading text for first enrollment (last in ordered array)
						  `Enrolled in ${_enrollment.site.name}`
						: `Changed enrollment to ${prettyAge(_enrollment.ageGroup)} in ${
								_enrollment.site.name
						  }`,
				body: `on ${dateFormatter(_enrollment.entry, false)}`,
				stepDate: _enrollment.entry,
				stepWeight: -1,
			})
		);
	}

	// enrollment stop step (only care about withdrawal for the last enrollment)
	if (enrollmentIndex === 0 && _enrollment.exit) {
		processStepProps.push(
			getSortableStep({
				heading: `Withdrawn from ${_enrollment.site.name}`,
				body: `on ${dateFormatter(_enrollment.exit, false)}`,
				stepDate: _enrollment.exit,
				stepWeight: 1,
			})
		);
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of fundings
 * @param fundings
 * @param enrollmentExit the value of Exit on the owning enrollment,
 * used to determine if switch to private pay step should be created
 */
export function fundingStepProps(
	fundings: DeepNonUndefineable<Funding[]> | null,
	enrollmentExit: Date | null
) {
	var processStepProps: SortableProcessStepProps[] = [];

	if (fundings) {
		fundings.sort(fundingStartSorter).forEach(funding => {
			// CDC Funding start if:
			// - source is CDC
			// - funding has first reporting period
			if (funding.source === FundingSource.CDC && funding.firstReportingPeriod) {
				var heading = `Switched funding to ${funding.source}`;
				var fundingTime = getFundingTime(funding);
				if (fundingTime) heading += ` - ${prettyFundingTime(fundingTime)}`;

				processStepProps.push(
					getSortableStep({
						heading,
						body: `on ${dateFormatter(funding.firstReportingPeriod.periodStart, false)}`,
						stepDate: funding.firstReportingPeriod.periodStart,
					})
				);
			}

			// Switch to private pay if:
			// - source is CDC
			// - funding has last reporting period
			// - enrollment is still ongoing (exit == null)
			// 	 OR last reporting period ends before enrollment exit
			// TODO: Handle other funding sources
			if (
				funding.source === FundingSource.CDC &&
				funding.lastReportingPeriod &&
				(!enrollmentExit || funding.lastReportingPeriod.periodEnd < enrollmentExit)
			) {
				var dayAfterLastReportingPeriodEnd = moment(funding.lastReportingPeriod.periodEnd)
					.add(1, 'days')
					.toDate();
				processStepProps.push(
					getSortableStep({
						heading: `Switched funding to ${NO_FUNDING}`,
						body: `on ${dateFormatter(dayAfterLastReportingPeriodEnd, false)}`,
						stepDate: dayAfterLastReportingPeriodEnd
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of family determinations
 * @param determinations
 */
export function determinationStepProps(
	determinations: DeepNonUndefineable<FamilyDetermination[]> | null
) {
	var processStepProps: SortableProcessStepProps[] = [];
	if (determinations) {
		determinations.forEach(determination => {
			if (!determination.notDisclosed && determination.determinationDate) {
				processStepProps.push(
					getSortableStep({
						heading: 'Income redetermined',
						body: `on ${dateFormatter(determination.determinationDate, false)}`,
						stepDate: determination.determinationDate,
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of C4KCertificates
 * @param c4KCertificates
 */
export function c4kCertificateStepProps(
	c4KCertificates: DeepNonUndefineable<C4KCertificate[]> | null
) {
	var processStepProps: SortableProcessStepProps[] = [];

	if (c4KCertificates) {
		c4KCertificates.sort(c4kCertificateSorter).forEach((c4kCertificate, idx) => {
			if (c4kCertificate.startDate) {
				processStepProps.push(
					getSortableStep({
						heading:
							idx === 0 ? 'Care 4 Kids certificate added' : 'Care 4 Kids certificate renewed',
						body: `on ${dateFormatter(c4kCertificate.startDate, false)}`,
						stepDate: c4kCertificate.startDate,
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Sorts SortableProcessStepProp objects inversely by sortDate, using sortWeight as a tie breaker
 * @param a
 * @param b
 */
export function enrollmentHistoryPropsSorter(
	a: SortableProcessStepProps,
	b: SortableProcessStepProps
) {
	var dateSorterRes = dateSorter(a.sortDate, b.sortDate, true);

	if (dateSorterRes !== 0) return dateSorterRes;

	if (a.sortWeight < b.sortWeight) return 1;
	if (a.sortWeight > b.sortWeight) return -1;
	return 0;
}

/**
 * Returns a SortableProcessStepProps object composed from the given parameters,
 * using the stepDate as sortDate, and to get correct value for isNew
 * @param heading
 * @param body
 * @param stepDate
 * @param stepWeight
 */
export function getSortableStep(params: {
	heading: string;
	body: string;
	stepDate: Date;
	stepWeight?: number;
}): SortableProcessStepProps {
	return {
		heading: params.heading,
		body: params.body,
		sortDate: params.stepDate,
		sortWeight: params.stepWeight ? params.stepWeight : 0,
		isNew: stepIsNew(params.stepDate),
	};
}

/**
 * Returns true is date is less than 1 week old, else false
 * @param stepDate
 */
export function stepIsNew(stepDate: Date) {
	return moment().add(-1, 'week') < moment(stepDate);
}
