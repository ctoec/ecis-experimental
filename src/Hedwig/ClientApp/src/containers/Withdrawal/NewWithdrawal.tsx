import React, { useContext, useReducer, useState, useEffect } from 'react';
import { History } from 'history';
import UserContext from '../../contexts/User/UserContext';
import {
	getIdForUser,
	validatePermissions,
	currentCdcFunding,
	currentC4kFunding,
	lastNReportingPeriods,
	reportingPeriodFormatter,
	enrollmentExitReasons,
} from '../../utils/models';
import useNewUseApi, { ApiError } from '../../hooks/newUseApi';
import { Enrollment, ReportingPeriod, Funding, ValidationProblemDetails } from '../../generated';
import { FormReducer, formReducer, updateData } from '../../utils/forms/form';
import { DeepNonUndefineable, DeepNonUndefineableArray } from '../../utils/types';
import ReportingPeriodContext from '../../contexts/ReportingPeriod/ReportingPeriodContext';
import {
	useFocusFirstError,
	hasValidationErrors,
	isBlockingValidationError,
	serverErrorForField,
} from '../../utils/validations';
import { childWithdrawnAlert, nameFormatter, splitCamelCase } from '../../utils/stringFormatters';
import AlertContext from '../../contexts/Alert/AlertContext';
import { validationErrorAlert } from '../../utils/stringFormatters/alertTextMakers';
import moment from 'moment';
import CommonContainer from '../CommonContainer';
import { InlineIcon, DateInput, ChoiceList, Button } from '../../components';
import { processBlockingValidationErrors } from '../../utils/validations/processBlockingValidationErrors';

type WithdrawalProps = {
	history: History;
	match: {
		params: {
			siteId: number;
			enrollmentId: number;
		};
	};
};

export default function NewWithdrawal({
	history,
	match: {
		params: { siteId, enrollmentId },
	},
}: WithdrawalProps) {
	const { user } = useContext(UserContext);
	const { setAlerts } = useContext(AlertContext);

	// set up initial form data
	const requestParams = {
		id: enrollmentId || 0,
		orgId: getIdForUser(user, 'org'),
		siteId: validatePermissions(user, 'site', siteId) ? siteId : 0,
	};

	const [getLoading, getError, getData] = useNewUseApi<Enrollment>(
		api =>
			api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet({
				...requestParams,
				include: ['child', 'family', 'determinations', 'fundings', 'sites'],
			}),
		{ skip: !enrollmentId || !user }
	);

	const [enrollment, updateEnrollment] = useReducer<FormReducer<DeepNonUndefineable<Enrollment>>>(
		formReducer,
		getData
	);

	useEffect(() => {
		// When get data has changed, update the enrollment to not be empty
		updateEnrollment(getData)
	}, [getData])

	// set up form variables and update functions
	const updateFormData = updateData<DeepNonUndefineable<Enrollment>>(updateEnrollment);

	const { exit: enrollmentEndDate, exitReason } = enrollment || {};
	const [lastReportingPeriod, setLastReportingPeriod] = useState<ReportingPeriod>();

	// set up supporting data contexts
	const { cdcReportingPeriods: reportingPeriods } = useContext(ReportingPeriodContext);
	const [reportingPeriodOptions, setReportingPeriodOptions] = useState(reportingPeriods);
	useEffect(() => {
		if (reportingPeriods) {
			setReportingPeriodOptions(
				lastNReportingPeriods(reportingPeriods, enrollmentEndDate || moment().toDate(), 5)
			);
		}
	}, [reportingPeriods, enrollmentEndDate]);

	// set up form state
	const [loading, setLoading] = useState(getLoading);
	const [error, setError] = useState<ApiError | null>(getError);
	const [hasAlertedOnError, setHasAlertedOnError] = useState(false);
	const [attempedSave, setAttempedSave] = useState(false);
	useFocusFirstError([error]);

	// set up convenience derived variables
	const isMissingInformation = hasValidationErrors(enrollment);
	const fundings =
		enrollment && enrollment.fundings
			? enrollment.fundings
			: ([] as DeepNonUndefineableArray<Funding>);
	const cdcFunding = currentCdcFunding(fundings);

	// set up PUT request to be triggered on save attempt
	const [putLoading, putError, putData] = useNewUseApi<Enrollment>(
		api =>
			api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPut({
				...requestParams,
				enrollment,
			}),
		{ skip: !attempedSave }
	);
	useEffect(() => {
		setLoading(putLoading);
		setError(putError);
		updateEnrollment(putData);
	}, [putLoading, putError, putData]);

	// handle PUT request
	useEffect(() => {
		if (!attempedSave) {
			return;
		}

		if (loading) {
			return;
		}

		if (!error) {
			setAlerts([childWithdrawnAlert(nameFormatter(enrollment.child))]);
			history.push(`/roster`);
			return;
		}

		if (!hasAlertedOnError) {
			if (!isBlockingValidationError(error)) {
				throw new Error(error.title || 'Unknown api error');
			}
			setAlerts([validationErrorAlert]);
		}
	}, [attempedSave, loading, error, enrollment, hasAlertedOnError, setAlerts]);

	// save function
	const save = () => {
		// handle enrollment updates that are not handled by form reducer updates
		let updatedFundings: Funding[] = fundings;
		if (cdcFunding && lastReportingPeriod) {
			updatedFundings = [
				...updatedFundings.filter(f => f.id !== cdcFunding.id),
				{
					...cdcFunding,
					lastReportingPeriodId: lastReportingPeriod.id,
				},
			];
		}

		// TODO: should we be doing this?
		// given that we're back to treating c4k certs as valid for one year,
		// I'm thinking not
		// const c4kFunding = currentC4kFunding(fundings);
		// if(c4kFunding) {
		// 	updatedFundings = [
		// 		...updatedFundings.filter(f => f.id != c4kFunding.id),
		// 		{
		// 			...c4kFunding,
		// 			certificateEndDate: enrollmentEndDate,
		// 		}
		// 	];
		// }

		enrollment.fundings = updatedFundings as DeepNonUndefineableArray<Funding>;
		setAttempedSave(true);
	};

	if (loading || !enrollment) {
		console.log('empty div');
		return <div className="Withdrawal"></div>;
	}

	return (
		<CommonContainer
			directionalLinkProps={{
				direction: 'left',
				to: `/roster/sites/${siteId}/enrollments/${enrollment.id}`,
				text: 'Back',
			}}
		>
			<div className="grid-container">
				<h1>Withdraw {nameFormatter(enrollment.child)}</h1>
				<div className="grid-row grid-gap oec-enrollment-details-section__content margin-top-1">
					<div className="mobile-lg:grid-col-6">
						<p>{enrollment.site.name}</p>
						<p>Age: {splitCamelCase(enrollment.ageGroup, '/')}</p>
						<p>Enrollment date: {enrollment.entry && enrollment.entry.toLocaleDateString()}</p>
					</div>
					{cdcFunding && (
						<div>
							<p>
								First reporting period:{' '}
								{cdcFunding.firstReportingPeriod
									? reportingPeriodFormatter(cdcFunding.firstReportingPeriod)
									: InlineIcon({ icon: 'attentionNeeded' })}
							</p>
						</div>
					)}
				</div>

				<div className="use-form mobile-lg:grid-col-12">
					<DateInput
						label="Enrollment end date"
						id="enrollment-end-date"
						name="exit"
						onChange={updateFormData(newDate => newDate.toDate())}
						date={enrollmentEndDate ? moment(enrollmentEndDate) : undefined}
						status={
							reportingPeriodOptions.length === 0
								? {
										type: 'error',
										id: 'last-reporting-period-error',
										message:
											'ECE Reporting only contains data for fiscal year 2020 and later. Please do not add children who withdrew prior to July 2019.',
								  }
								: serverErrorForField(
										hasAlertedOnError,
										setHasAlertedOnError,
										'exit',
										error,
										'This information is required for withdrawal'
								  )
						}
					/>
					<ChoiceList
						type="select"
						label="Reason"
						id="exit-reason"
						options={Object.entries(enrollmentExitReasons).map(([key, reason]) => ({
							value: key,
							text: reason,
						}))}
						selected={exitReason ? [exitReason] : undefined}
						otherInputLabel="Other"
						name="exitReason"
						onChange={updateFormData()}
						status={serverErrorForField(
							hasAlertedOnError,
							setHasAlertedOnError,
							'exitReason',
							error,
							'This information is required for withdrawal'
						)}
					/>
					{cdcFunding && (
						<ChoiceList
							type="select"
							label="Last reporting period"
							id="last-reporting-period"
							options={reportingPeriodOptions.map(period => ({
								value: '' + period.id,
								text: reportingPeriodFormatter(period, { extended: true }),
							}))}
							onChange={event => {
								const chosen = reportingPeriods.find<ReportingPeriod>(
									period => period.id === parseInt(event.target.value)
								);
								setLastReportingPeriod(chosen);
							}}
							status={serverErrorForField(
								hasAlertedOnError,
								setHasAlertedOnError,
								'fundings',
								error,
								// if last reporting period exists, then the value is invalid
								// so do not display required information copy
								lastReportingPeriod ? undefined : 'This information is required for withdrawal'
							)}
						/>
					)}
				</div>

				<div className="grid-row margin-y-6">
					<div className="mobile-lg:grid-col-auto">
						<Button
							text="Cancel"
							href={`/roster/sites/${siteId}/enrollments/${enrollment.id}/`}
							appearance="outline"
						/>
					</div>
					<div className="mobile-lg:grid-col-auto padding-left-2">
						<Button text="Confirm and withdraw" onClick={save} disabled={isMissingInformation} />
					</div>
				</div>
			</div>
		</CommonContainer>
	);
}
