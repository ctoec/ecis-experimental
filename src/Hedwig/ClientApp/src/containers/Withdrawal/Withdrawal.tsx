import React, { useContext, useState, useEffect } from "react";
import { History } from 'history';
import UserContext from "../../contexts/User/UserContext";
import { Enrollment, Funding, FundingSource, ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest, ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPutRequest, ValidationProblemDetails, ValidationProblemDetailsFromJSON, ReportingPeriod } from "../../generated";
import nameFormatter from "../../utils/nameFormatter";
import { DeepNonUndefineable, tsFilter } from "../../utils/types";
import { generateFundingTag, enrollmentExitReasons, currentFunding } from "../../utils/models";
import DatePicker from "../../components/DatePicker/DatePicker";
import Dropdown from "../../components/Dropdown/Dropdown";
import useApi from "../../hooks/useApi";
import { nextNReportingPeriods, lastNReportingPeriods } from "../../utils/models/reportingPeriod";
import getIdForUser from "../../utils/getIdForUser";
import Button from "../../components/Button/Button";
import CommonContainer from "../CommonContainer";
import { clientErrorForField, serverErrorForField } from "../../utils/validations";
import ReportingPeriodContext from "../../contexts/ReportingPeriod/ReportingPeriodContext";
import { processBlockingValidationErrors } from "../../utils/validations/processBlockingValidationErrors";
import InlineIcon from "../../components/InlineIcon/InlineIcon";
import { AlertProps } from "../../components/Alert/Alert";
import AlertContext from "../../contexts/Alert/AlertContext";

type WithdrawalProps = {
  history: History,
  match: {
    params: {
      enrollmentId?: number;
    }
  }
};

export default function Withdrawal({ 
  history,
  match: {
    params: { enrollmentId }
  }
}: WithdrawalProps) {
  const { user } = useContext(UserContext);
  const { getAlerts, setAlerts } = useContext(AlertContext);
  const { cdcReportingPeriods: reportingPeriods } = useContext(ReportingPeriodContext);

  const defaultParams: ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest = {
    id: enrollmentId || 0,
    orgId: getIdForUser(user, 'org'),
    siteId: getIdForUser(user, 'site'),
  }

  const [loading, error, enrollment, mutate] = useApi<Enrollment>(
    api => api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet({
      ...defaultParams,
      include: ['child', 'family', 'determinations', 'fundings', 'sites']
    }),
    [enrollmentId, user],
    {
      skip: !enrollmentId
    }
  );

  const [enrollmentEndDate, updateEnrollmentEndDate] = useState<Date>();
  const [exitReason, updateExitReason] = useState<string>();
  const [lastReportingPeriod, updateLastReportingPeriod] = useState<ReportingPeriod>();

  const [attemptedSave, setAttemptedSave] = useState(false);
  const [apiError, setApiError] = useState<ValidationProblemDetails>();

  if(loading || error || !enrollment) {
    return <div className="Withdrawl"></div>;
  }

  const fundings = enrollment.fundings ? enrollment.fundings : [] as DeepNonUndefineable<Funding[]>;
  const cdcFundings = fundings.filter(funding => funding.source === FundingSource.CDC);
  const cdcFunding = currentFunding(cdcFundings);

  const save = () => {
    setAttemptedSave(true);

    // funded enrollments without first reporting period are not valid for withdrawl
    // TODO add option to add first reporting period here
    if(cdcFunding && !cdcFunding.firstReportingPeriod) {
      setAlerts([{
        type: 'error',
        heading: 'Cannot withdraw',
        text: 'CDC funded enrollments must have first reporting period for withdrawal'
      }]);

      return;
    }

    //enrollment end date (exit) is required for withdrawl
    if(!enrollmentEndDate) {
      return;
    }

    let updatedFundings: Funding[] = fundings;
    if(cdcFunding && lastReportingPeriod) {
      updatedFundings = [
        ...(fundings.filter<DeepNonUndefineable<Funding>>(funding => funding.id !== cdcFunding.id)),
        {
          ...cdcFunding,
          lastReportingPeriodId: lastReportingPeriod.id,
          lastReportingPeriod: lastReportingPeriod
        }
      ]
    }
    const putParams: ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPutRequest = {
      ...defaultParams,
      enrollment: {
        ...enrollment,
        exit: enrollmentEndDate,
        exitReason,
        fundings: updatedFundings
      }
    }

    mutate(api => api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPut(putParams))
      .then(() => {
        history.push(`/roster`)
      })
      .catch(async (error) => {
        setApiError(ValidationProblemDetailsFromJSON(error));
      });
  }

  return (
    <CommonContainer
      directionalLinkProps={
        {direction: 'left', to: `/roster/enrollments/${enrollment.id}`, text:'Back'}
      }
    >
        <section className="grid-container">
          <h1>Withdraw from program</h1>
          <div className="margin-left-5">
            <h2>{nameFormatter(enrollment.child)}</h2>
            <div className="grid-row grid-gap">
              <div className="mobile-lg:grid-col-6">
                <h4>Current enrollment</h4>
                <p>{enrollment.site.name}</p>
                <p>Age: {enrollment.ageGroup}</p>
                <p>Enrollment date: {enrollment.entry && enrollment.entry.toLocaleDateString()}</p>              </div>
              {cdcFunding && 
                <div className="mobile-lg:grid-col-6">
                  <h4>Funding</h4>
                  {generateFundingTag(cdcFunding)}
                  <p>Enrollment: {cdcFunding.time}</p>
                  <p>First reporting period: {cdcFunding.firstReportingPeriod 
                    ? cdcFunding.firstReportingPeriod.period.toLocaleDateString()
                    : InlineIcon({icon: 'attentionNeeded'})
                  }</p>
                </div>
              }
            </div>
          
            <div className="usa-form">
              <div className="grid-row grid-gap">
                <div className="mobile-lg:grid-col-12">
                  <DatePicker
                    label="Enrollment end date"
                    id="enrollment-end-date"
                    onChange={range => 
                      updateEnrollmentEndDate((range.startDate && range.startDate.toDate()) || undefined)
                    }
                    dateRange={{ startDate: null, endDate: null}}
                    status={(apiError && processBlockingValidationErrors("exit", apiError.errors))
                    ? serverErrorForField(
                      "exit",
                      apiError
                    )
                    : clientErrorForField(
                      "exit",
                      enrollmentEndDate,
                      attemptedSave,
                      "This information is required for withdrawal"
                    )}
                  />
                  <Dropdown
                    label="Reason"
                    id="exit-reason"
                    options={Object.entries(enrollmentExitReasons).map(
                      ([key, reason]) => ({value: key, text: reason})
                    )}
                    otherText="Other"
                    noSelectionText="-Select-"
                    onChange={event => updateExitReason(event.target.value)}
                    status={serverErrorForField(
                      "exitreason",
                      apiError,
                      "This information is required for withdrawal"
                    )}
                  />
                  <Dropdown
                    label="Last reporting period"
                    id="last-reporting-period"
                    options={
                        lastNReportingPeriods(
                        reportingPeriods,
                        enrollmentEndDate || new Date(Date.now()),
                        5
                      ).map(period => 
                        ({ 
                          value: '' + period.id,
                          text: `${period.periodStart.toLocaleDateString()} - ${period.periodEnd.toLocaleDateString()}`
                        })
                      )
                    }
                    noSelectionText="-Select-"
                    onChange={event => {
                      const chosen = reportingPeriods.find<ReportingPeriod>( period => period.id === parseInt(event.target.value))
                      updateLastReportingPeriod(chosen);
                    }}
                    status={serverErrorForField(
                      "fundings",
                      apiError,
                      lastReportingPeriod ? undefined : 'This information is required for withdrawal'
                    )}
                  />
                </div>
              </div>
              <div className="grid-row">
                <div className="mobile-lg:grid-col-auto">
                  <Button text="Cancel" href={`/roster/enrollments/${enrollment.id}/`} appearance='base' />
                </div>
                <div className="mobile-lg:grid-col-auto">
                  <Button text="Confirm and withdraw" onClick={save} />
                </div>
              </div>
            </div>
          </div>
        </section>
    </CommonContainer>
  )
}