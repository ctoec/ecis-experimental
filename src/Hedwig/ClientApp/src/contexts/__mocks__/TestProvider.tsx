import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, History } from 'history';
import AuthenticationContext from '../Authentication/AuthenticationContext';
import UserContext from '../User/UserContext';
import ReportingPeriodContext from '../ReportingPeriod/ReportingPeriodContext';
import { User, ReportingPeriod } from '../../generated';
import { cdcReportingPeriods as _cdcReportingPeriods, user as _user } from '../../tests/data';

type TestProviderProps = {
	user?: User;
	cdcReportingPeriods?: ReportingPeriod[];
	history?: History<any>;
};

const TestProvider: React.FC<TestProviderProps> = ({
	children,
	user = _user,
	cdcReportingPeriods = _cdcReportingPeriods,
	history = createMemoryHistory(),
}) => {
	return (
		<AuthenticationContext.Provider
			value={{ accessToken: '', withFreshToken: jest.fn(), loading: false }}
		>
			<UserContext.Provider value={{ user, loading: false }}>
				<ReportingPeriodContext.Provider value={{ cdcReportingPeriods }}>
					<Router history={history}>{children}</Router>
				</ReportingPeriodContext.Provider>
			</UserContext.Provider>
		</AuthenticationContext.Provider>
	);
};

export default TestProvider;
