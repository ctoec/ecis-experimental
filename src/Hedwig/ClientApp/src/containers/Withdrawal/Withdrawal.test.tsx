// Variables used in jest mockes -- must start with `mock`
import { mockAllFakeEnrollments, mockCompleteEnrollment } from '../../tests/data';
import mockUseApi, {
	mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet,
	mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPut,
} from '../../hooks/__mocks__/newUseApi';
import { accessibilityTestHelper } from '../../tests/helpers';
import TestProvider from '../../contexts/__mocks__/TestProvider';

// Jest mocks must occur before later imports
jest.mock('../../hooks/useApi', () => {
	return mockUseApi({
		apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet: (_: any) =>
			mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(mockAllFakeEnrollments)(_),
		apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPut: (_: any) =>
			mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPut(mockAllFakeEnrollments)(_),
	});
});

let mockSiteId: number | undefined = 1;
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
	useParams: () => ({
		id: mockSiteId,
	}),
}));

import React from 'react';
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import 'react-dates/initialize';
import mockdate from 'mockdate';
import Withdrawal from './Withdrawal';

const fakeDate = '2019-09-30';

beforeAll(() => {
	mockdate.set(fakeDate);
});

afterAll(() => {
	mockdate.reset();
	jest.resetModules();
});

describe('Withdrawal', () => {
	it('matches snapshot', () => {
		const history = createMemoryHistory();
		const { asFragment } = render(
			<TestProvider>
				<Withdrawal
					history={history}
					match={{
						params: {
							siteId: mockCompleteEnrollment.siteId,
							enrollmentId: mockCompleteEnrollment.id,
						},
					}}
				/>
			</TestProvider>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	accessibilityTestHelper(
		<TestProvider>
			<Withdrawal
				history={createMemoryHistory()}
				match={{
					params: {
						siteId: mockCompleteEnrollment.siteId,
						enrollmentId: mockCompleteEnrollment.id,
					},
				}}
			/>
		</TestProvider>
	);
});
