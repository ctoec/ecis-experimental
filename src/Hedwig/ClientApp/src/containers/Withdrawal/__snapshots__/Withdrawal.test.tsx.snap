// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Withdrawal matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Withdrawal"
  />
</DocumentFragment>
`;
